website/blog/2024/12-13-scrape-google-maps-using-python/index.md

```diff
@@ -0,0 +1,481 @@
+
+In the code:
+
+- `query_selector`: Returns first DOM element matching CSS selector, useful for single items like a name or rating
+- `query_selector_all`: Returns all matching elements, ideal for multiple items like amenities
+- `inner_text()`: Extracts text content
+- Some hotels might not have all the information available - we handle this with 'N/A’
+
+When you run this script, you'll see output similar to this:
+
+```json
+{
+    "name": "GRAND KALINGA HOTEL",
+    "rating": "4.2",
+    "reviews": "1,171",
+    "price": "\u20b91,760",
+    "link": "https://www.google.com/maps/place/GRAND+KALINGA+HOTEL/data=!4m10!3m9!1s0x3bae160e0ce07789:0xb15bf736f4238e6a!5m2!4m1!1i2!8m2!3d12.9762259!4d77.5786043!16s%2Fg%2F11sp32pz28!19sChIJiXfgDA4WrjsRao4j9Db3W7E?authuser=0&hl=en&rclk=1",
+    "amenities": [
+        "Pool available",
+        "Free breakfast available",
+        "Free Wi-Fi available",
+        "Free parking available"
+    ]
+}
+```
+
+### 5. Scroll to load more
+
+When scraping Google Maps, you'll notice that not all results load at once. Let's handle this infinite scroll pagination!
+
+First, we need a function that can handle the scrolling and detect when we've hit the bottom. Copy-paste this new function in the `gmap_scraper.py` file:
+
+```python
+async def load_more_items(page) -> bool:
+    # Locate the scrollable feed container
+    feed = await page.query_selector('div[role="feed"]')
+    if not feed:
+        return False
+    
+    # Get the current scroll position
+    prev_scroll = await feed.evaluate("(element) => element.scrollTop")
+
+    # Scroll down to load more items
+    await feed.evaluate("(element) => element.scrollTop += 800")
+    await page.wait_for_timeout(2000)  # Allow content to load
+
+    # Check if the page actually scrolled
+    new_scroll = await feed.evaluate("(element) => element.scrollTop")
+    if new_scroll <= prev_scroll:  # No further scroll means end of listings
+        return False
+    
+    # Extra wait for dynamic content to appear
+    await page.wait_for_timeout(1000)
+    return True
+```
+
+Run this code using:
+
+```bash
+$ python3 gmap_scraper.py
+```
+
+You should see an output like this:
+
+![scrape-google-maps-with-crawlee-screenshot-handle-pagination.png](https://prod-files-secure.s3.us-west-2.amazonaws.com/63b6fe41-a503-4e50-9b1e-caa1e011ae25/cba448f3-f260-4cb4-b068-93ca4a9d54de/d905f2ab-d448-460d-b8eb-f68d2b2f842d.png)
+
+### 6. Exporting Google Maps data to JSON
+
+Once you've successfully scraped data from Google Maps, it's important to save that data in a format that is both accessible and easy to work with. JSON is an excellent choice for this purpose.
+
+Here's a simple code that saves your scraped data to a JSON file:
+
+```python
+import json
+
+# Your scraped data will be stored in this list
+all_data = [
+    ...
+]
+
+# Save data to JSON file
+with open('google_maps_data.json', 'w', encoding='utf-8') as f:
+    json.dump(all_data, f, ensure_ascii=False, indent=2)
```

Comment:
If you use the default dataset for this, you can simply do `crawler.export_data_json('path', ensure_ascii=False, indent=2)`